huaweicloud/services/ces/resource_huaweicloud_ces_alarmrule.go

```diff
@@ -394,69 +531,275 @@ func resourceAlarmRuleRead(_ context.Context, d *schema.ResourceData, meta inter
+	clientV1, err := config.CesV1Client(config.GetRegion(d))
 	if err != nil {
-		return fmtp.DiagErrorf("Error creating Cloud Eye Service client: %s", err)
+		return diag.Errorf("error creating Cloud Eye Service v1 client: %s", err)
+	}
+	clientV2, err := config.CesV2Client(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating Cloud Eye Service v2 client: %s", err)
 	}
 
 	arId := d.Id()
 
-	if d.HasChange("alarm_enabled") {
-		enabled := d.Get("alarm_enabled").(bool)
-		enableOpts := alarmrule.EnableOpts{
-			AlarmEnabled: enabled,
+	if d.HasChanges("alarm_name", "alarm_description", "alarm_action_enabled", "alarm_actions", "ok_actions") {
+		updateOpts := alarmrule.UpdateOpts{
+			Name:         d.Get("alarm_name").(string),
+			AlarmActions: buildAlarmAction(d, "alarm_actions"),
+			OkActions:    buildAlarmAction(d, "ok_actions"),
+		}
+
+		description := d.Get("alarm_description").(string)
+		updateOpts.Description = &description
+
+		// add alarm_action_enabled to the updateOpts only when it's changed
+		// this can avoid API error
+		if d.HasChange("alarm_action_enabled") {
+			actionEnabled := d.Get("alarm_action_enabled").(bool)
+			updateOpts.ActionEnabled = &actionEnabled
 		}
-		logp.Printf("[DEBUG] Updating %s %s to %#v", nameCESAR, arId, enabled)
 
-		timeout := d.Timeout(schema.TimeoutUpdate)
-		//lintignore:R006
-		err = resource.RetryContext(ctx, timeout, func() *resource.RetryError {
-			err := alarmrule.Enable(client, arId, enableOpts).ExtractErr()
+		log.Printf("[DEBUG] Updating %s %s opts: %#v", nameCESAR, arId, updateOpts)
+		err := alarmrule.Update(clientV1, arId, updateOpts).ExtractErr()
+		if err != nil {
+			return diag.Errorf("error updating %s %s: %s", nameCESAR, arId, err)
+		}
+	}
+
+	if d.HasChange("metric.0.dimensions") {
+		oldDimensions, newDimensions := d.GetChange("metric.0.dimensions")
+
+		if len(oldDimensions.([]interface{})) > 0 {
+			updateDimensionsOpts := alarmrulev2.UpdateResourcesOpts{
+				Resources: buildDimensionsOpts(oldDimensions.([]interface{})),
+			}
+
+			err := alarmrulev2.BatchResources(clientV2, arId, "batch-delete", updateDimensionsOpts).ExtractErr()
 			if err != nil {
-				return common.CheckForRetryableError(err)
+				return diag.Errorf("error deleting old dimensions of %s %s: %s", nameCESAR, arId, err)
 			}
-			return nil
-		})
+		}
+
+		if len(newDimensions.([]interface{})) > 0 {
+			updateDimensionsOpts := alarmrulev2.UpdateResourcesOpts{
+				Resources: buildDimensionsOpts(newDimensions.([]interface{})),
+			}
+			err := alarmrulev2.BatchResources(clientV2, arId, "batch-create", updateDimensionsOpts).ExtractErr()
+			if err != nil {
+				return diag.Errorf("error creating new dimensions of %s %s: %s", nameCESAR, arId, err)
+			}
+		}
+	}
+
+	// update conditon if alarm_level changed
+	if d.HasChange("alarm_level") {
+		level := 2
+		if v, ok := d.GetOk("alarm_level"); ok {
+			level = v.(int)
+		}
+
+		_, _, metricName := buildResourcesOpts(d)
```

Comment:
we can define the global metricName outside if d.HasChanges()...

Reply:
fixed